Have system use static buckets and expose API

muh200 opened the pull request.

Commit:
Created Dockerfile for gcsfuse container.

Separated the Dockerfile for the ffmpeg-worker into two Dockerfiles: one
container for ffmpeg and one for gcsfuse.

Commit:
Created k8s deployment file for ffmpeg worker.

Commit:
Created k8s daemonset file for gcsfuse daemonset.

Commit:
Changed worker to use statically mounted buckets.

The worker is not responsible for mounting and unmounting the bucket.
The buckets are statically mounted by the gcsfuse daemonset.

Commit:
Implemented server returning FFmpeg exit code.

Commit:
Added configuration for Cloud Endpoints API.

Now, every worker pod has two containers: one for the API frontend and
one for the API backend.
There is now a public FFmpeg service that runs on port 80.

Commit:
Implemented client to specify IP and API key

The ip address and optional port number are specified as arguments on the
command line. The API key is stored in an environment variable.

Commit:
Created a kustomization file.

This would allow users to use this file as a base and customize the
deployment to their own preferences.

Commit:
Wrote documentation about deploying to GKE.

isasi commented on docs/deployment.md:
```diff
@@ -0,0 +1,99 @@
+
+1. Create a Google Cloud Project if you do not have one. Enable the Cloud Build and Container Registry APIs.
+
+2. Set the project and compute zone.
```
Probably add "Open a new Cloud Shell in the Cloud console of the project " before listing the shell commands

isasi commented on docs/deployment.md:
````diff
@@ -0,0 +1,99 @@
+gcloud config set compute/zone $COMPUTE_ZONE
+```
+
+3. Create a GKE cluster with [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) enabled. Replace `$CLUSTER_NAME` and `$PROJECT_ID` with the name of the new cluster and the project ID respectively.
````
Can you  group these individual instructions into higher level objectives . For example Point 4, 5 can be grouped as 4. Setting up GCSFuse for the project 
      i) Create serivce account ...
      ii) Add policy binding ...
etc.

isasi commented on gcsfuse-daemonset/gcsfuse-daemonset.yaml:
```diff
@@ -0,0 +1,40 @@
+        # Replace $BUCKET_NAME with the name of the bucket to mount.
+        command: ["/bin/bash", "-c", "mkdir -p /buckets/$BUCKET_NAME/ && gcsfuse -o allow_other --uid 999 --foreground $BUCKET_NAME /buckets/$BUCKET_NAME/"]
+        lifecycle:
+          preStop:
```
can you add in the comment above what this command does and what -o allow_other and --uid 999 do

isasi commented on worker/api_config.yaml:
```diff
@@ -0,0 +1,9 @@
+# Below, replace MY_PROJECT_ID with your Google Cloud Project ID.
```
GCLOUD_PROJECT

Commit:
Added instructions to open Cloud Shell.

Commit:
Added instruction to replace parameters in command

Commit:
Grouped similar subtasks into higher level tasks.

Commit:
Added comments explaining gcsfuse command.

Commit:
Fixed typo from MY_PROJECT_ID to $GCLOUD_PROJECT.

Commit:
Returned FFmpeg process resource usage to client.

Commit:
fixup! Added configuration for Cloud Endpoints API.

isasi commented on worker/ffmpeg_worker.py:
```diff
@@ -43,53 +47,50 @@ def transcode(self, request: FFmpegRequest, context) -> FFmpegResponse:
+    def __init__(self, args):
+        self._pipe = subprocess.Popen(args,
+                                      stdout=subprocess.PIPE,
+                                      stderr=subprocess.STDOUT,
```
this var should be named self._subprocess or self._ffmpeg_sp

isasi commented on worker/ffmpeg_worker.py:
```diff
@@ -43,53 +47,50 @@ def transcode(self, request: FFmpegRequest, context) -> FFmpegResponse:
+    def __iter__(self):
+        yield from self._pipe.stdout
+        _, status, self.rusage = os.wait4(self._pipe.pid, 0)
+        self.returncode = (os.WEXITSTATUS(status) if os.WIFEXITED(status)
```
just return the status directly and let client handle the semantics of dealing with whether it was exited or signaled .

isasi commented on worker/client.py:
```diff
@@ -15,12 +15,14 @@
 
 The following is a sample command:
-python3 client.py my-bucket-1 my-bucket-2 -- -i my-bucket-1/input.mp4 my-bucket-2/output.avi
+python3 client.py -i my-bucket-1/input.mp4 my-bucket-2/output.avi
```
can you add the server ip here in the example

isasi commented on worker/client.py:
```diff
@@ -29,17 +31,28 @@
+            print(response.log_line, end='')
+        else:
+            print(f'Exited with code {response.exit_status.exit_code}')
+
```
Also print the Resource usage proto maybe

Commit:
Changed exit_code to be status returned by wait().

Now the client is reponsible for parsing the exit status and determining
if the process was killed by a signal or not.

Commit:
Modified client to print resource usage of FFmpeg.

Commit:
Changed member _pipe of Process to _subprocess.

Commit:
Added server IP to example usage in comments.

Commit:
Removed extra new lines and broke long lines.

Commit:
Disabled gcsfuse caching to enforce consistency.

Since gcsfuse is mounted on multiple nodes, the file cache could become
inconsistent across the nodes. This is why it is necessary to disable caching.
